File: dotnet/ImperativeVsFunctional/FunctionalChallenges.Tests/SortingAlgorithmsTests.cs
using static Challenges.SortingAlgorithms;

namespace Challenges.Tests;

public class SortingAlgorithmsTests
{
    [Theory]
    [InlineData(new [] { 5, 3, 4, 2, 1 }, new [] { 1, 2, 3, 4, 5 })]
    [InlineData(new [] { 2, 2, 3, 1, 1 }, new [] { 1, 1, 2, 2, 3 })]
    [InlineData(new [] { 1, 2, 3, 4, 5 }, new [] { 1, 2, 3, 4, 5})]
    public void BubbleSort_ReturnsSortedArray(int[] numbers, int[] expected)
    {
        var actual = BubbleSort(numbers);
        
        Assert.Equal(expected, actual);
    }
    
    [Theory]
    [InlineData(new [] { 5, 3, 4, 2, 1 }, new [] { 1, 2, 3, 4, 5 })]
    [InlineData(new [] { 2, 2, 3, 1, 1 }, new [] { 1, 1, 2, 2, 3 })]
    [InlineData(new [] { 1, 2, 3, 4, 5 }, new [] { 1, 2, 3, 4, 5})] 
    public void InsertionSort_ReturnsSortedArray(int[] numbers, int[] expected)
    {
        var actual = InsertionSort(numbers);
        
        Assert.Equal(expected, actual);
    }
    
    [Theory]
    [InlineData(new [] { 5, 3, 4, 2, 1 }, new [] { 1, 2, 3, 4, 5 })]
    [InlineData(new [] { 2, 2, 3, 1, 1 }, new [] { 1, 1, 2, 2, 3 })]
    [InlineData(new [] { 1, 2, 3, 4, 5 }, new [] { 1, 2, 3, 4, 5})] 
    public void SelectionSort_ReturnsSortedArray(int[] numbers, int[] expected)
    {
        var actual = SelectionSort(numbers);
        
        Assert.Equal(expected, actual);
    }
    
    [Theory]
    [InlineData(new [] { 5, 3, 4, 2, 1 }, new [] { 1, 2, 3, 4, 5 })]
    [InlineData(new [] { 2, 2, 3, 1, 1 }, new [] { 1, 1, 2, 2, 3 })]
    [InlineData(new [] { 1, 2, 3, 4, 5 }, new [] { 1, 2, 3, 4, 5})] 
    public void CocktailSort_ReturnsSortedArray(int[] numbers, int[] expected)
    {
        var actual = CocktailSort(numbers);
        
        Assert.Equal(expected, actual);
    }
    
    [Theory]
    [InlineData(new [] { 5, 3, 4, 2, 1 }, new [] { 1, 2, 3, 4, 5 })]
    [InlineData(new [] { 2, 2, 3, 1, 1 }, new [] { 1, 1, 2, 2, 3 })]
    [InlineData(new [] { 1, 2, 3, 4, 5 }, new [] { 1, 2, 3, 4, 5})] 
    public void ShellSort_ReturnsSortedArray(int[] numbers, int[] expected)
    {
        var actual = ShellSort(numbers);
        
        Assert.Equal(expected, actual);
    }
    
    [Theory]
    [InlineData(new [] { 5, 3, 4, 2, 1 }, new [] { 1, 2, 3, 4, 5 })]
    [InlineData(new [] { 2, 2, 3, 1, 1 }, new [] { 1, 1, 2, 2, 3 })]
    [InlineData(new [] { 1, 2, 3, 4, 5 }, new [] { 1, 2, 3, 4, 5})] 
    public void QuickSort_ReturnsSortedArray(int[] numbers, int[] expected)
    {
        var actual = QuickSort(numbers);
        
        Assert.Equal(expected, actual);
    }
}

File: typescript/challenges/test/sorting-algorithms-tests.ts
import {bubbleSort, cocktailSort, insertionSort, quickSort, selectionSort, shellSort} from "../src/sorting-algorithms";

test.each([
    [[5, 3, 4, 2, 1], [1, 2, 3, 4, 5]],
    [[2, 2, 3, 1, 1], [1, 1, 2, 2, 3]],
    [[1, 2, 3, 4, 5], [1, 2, 3, 4, 5]]
])('BubbleSort returns sorted array', (listOfNumbers, expected) => {
    expect(bubbleSort(listOfNumbers)).toEqual(expected);
});

test.each([
    [[5, 3, 4, 2, 1], [1, 2, 3, 4, 5]],
    [[2, 2, 3, 1, 1], [1, 1, 2, 2, 3]],
    [[1, 2, 3, 4, 5], [1, 2, 3, 4, 5]]
])('InsertionSort returns sorted array', (listOfNumbers, expected) => {
    expect(insertionSort(listOfNumbers)).toEqual(expected);
});

test.each([
    [[5, 3, 4, 2, 1], [1, 2, 3, 4, 5]],
    [[2, 2, 3, 1, 1], [1, 1, 2, 2, 3]],
    [[1, 2, 3, 4, 5], [1, 2, 3, 4, 5]]
])('SelectionSort returns sorted array', (listOfNumbers, expected) => {
    expect(selectionSort(listOfNumbers)).toEqual(expected);
});

test.each([
    [[5, 3, 4, 2, 1], [1, 2, 3, 4, 5]],
    [[2, 2, 3, 1, 1], [1, 1, 2, 2, 3]],
    [[1, 2, 3, 4, 5], [1, 2, 3, 4, 5]]
])('CocktailSort returns sorted array', (listOfNumbers, expected) => {
    expect(cocktailSort(listOfNumbers)).toEqual(expected);
});

test.each([
    [[5, 3, 4, 2, 1], [1, 2, 3, 4, 5]],
    [[2, 2, 3, 1, 1], [1, 1, 2, 2, 3]],
    [[1, 2, 3, 4, 5], [1, 2, 3, 4, 5]]
])('ShellSort returns sorted array', (listOfNumbers, expected) => {
    expect(shellSort(listOfNumbers)).toEqual(expected);
});

test.each([
    [[5, 3, 4, 2, 1], [1, 2, 3, 4, 5]],
    [[2, 2, 3, 1, 1], [1, 1, 2, 2, 3]],
    [[1, 2, 3, 4, 5], [1, 2, 3, 4, 5]]
])('QuickSort returns sorted array', (listOfNumbers, expected) => {
    expect(quickSort(listOfNumbers)).toEqual(expected);
});
